src/iterator.rs

```diff
@@ -49,61 +46,390 @@
-use std::sync::atomic::{AtomicBool, Ordering};
-use std::sync::{Arc, Mutex};
+/// A backend module containing generic types which can be used to implement
+/// iterator or stream style signal handling using different IO types for
+/// the internal self-pipe.
+pub mod backend {
+    use std::borrow::Borrow;
+    use std::io::Error;
+    use std::os::unix::io::AsRawFd;
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use std::sync::{Arc, Mutex};
 
-use libc::{self, c_int};
+    use libc::{self, c_int};
 
-use crate::SigId;
+    use crate::SigId;
 
-/// Maximal signal number we support.
-const MAX_SIGNUM: usize = 128;
+    /// Maximal signal number we support.
+    const MAX_SIGNUM: usize = 128;
 
-#[derive(Debug)]
-struct Waker {
-    pending: Vec<AtomicBool>,
-    closed: AtomicBool,
-    read: UnixStream,
-    write: UnixStream,
-}
+    #[derive(Debug)]
+    struct Waker<R, W> {
+        pending: Arc<Vec<AtomicBool>>,
+        closed: AtomicBool,
+        read: R,
+        write: W,
+    }
+
+    impl<R, W: AsRawFd> Waker<R, W> {
+        /// Create a new instance for the given read and write pipe ends.
+        fn new(read: R, write: W) -> Self {
+            let pending = Arc::new((0..MAX_SIGNUM).map(|_| AtomicBool::new(false)).collect());
+
+            Self {
+                pending,
+                closed: AtomicBool::new(false),
+                read,
+                write,
+            }
+        }
+
+        /// Get a [`Pending`](struct.Pending) instance for iterating through recieved signals.
+        fn pending(&self) -> Pending {
+            Pending::new(Arc::clone(&self.pending))
+        }
+
+        /// Sends a wakeup signal to the internal wakeup pipe.
+        fn wake(&self) {
+            unsafe {
+                // See the comment at pipe::write.
+                //
+                // We don't use pipe::write, because it expects the FD to be already in non-blocking
+                // mode. That's because it needs to support actual pipes. We can afford send here,
+                // which has flags.
+                libc::send(
+                    self.write.as_raw_fd(),
+                    b"X" as *const _ as *const _,
+                    1,
+                    libc::MSG_DONTWAIT,
+                );
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    struct RegisteredSignals(Mutex<Vec<Option<SigId>>>);
+
+    impl Drop for RegisteredSignals {
+        fn drop(&mut self) {
+            let lock = self.0.lock().unwrap();
+            for id in lock.iter().filter_map(|s| *s) {
+                crate::unregister(id);
+            }
+        }
+    }
+
+    /// A struct for delivering received signals to the main program flow.
+    /// The self-pipe IO type is generic. See the [with_pipe](#method.with_pipe)
+    /// method for requirements for the IO type.
+    #[derive(Debug)]
+    pub struct SignalDelivery<R, W> {
+        ids: Arc<RegisteredSignals>,
+        waker: Arc<Waker<R, W>>,
+    }
```

Comment:
I think you're right, this would lead to some cycle => Drop on RegisteredSignals will only be called when all signals are unregistered but unregistering signals happens only in the drop implementation of RegisteredSignals.